### ISSUE TITLE
LSRL aggregates depth scores into one trajectory advantage and does not make one API call per depth

### ISSUE THREAD

**Issue by microprediction** [microprediction/latentreasoning#72]
## Reviewed source

Commit: [`11b700ba0c9cc03b0e1a54db15fe06d15fad8e51`](https://github.com/microprediction/latentreasoning/commit/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51)

- [`training.html` lines 471–487](https://github.com/microprediction/latentreasoning/blob/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51/training.html#L471-L487) defines a process reward as assigning credit to the intermediate step that earned it, then says LSRL supplies a “dense per-depth reward inside GRPO” and costs “an API call per depth per rollout.”
- [`bibliography.html` lines 1139–1141](https://github.com/microprediction/latentreasoning/blob/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51/bibliography.html#L1139-L1141) repeats the dense-per-depth characterization.

## What the paper actually optimizes

Ren’s [LSRL paper](https://aclanthology.org/2025.findings-emnlp.669.pdf) does obtain a grader score `R_k` for every decoded recurrent depth, but Eqs. (4)–(8) do **not** attach a separate policy-gradient term to depth `k`.

First, all depth scores are collapsed into a single discounted process scalar:

```text
R_proc = sum_{k=1}^r gamma^(k-1) R_k
```

That scalar is then mixed with terminal correctness:

```text
R_tot = 0.7 * 1[final answer correct] + 0.3 * R_proc
```

Equation (7) forms one group-relative trajectory advantage `A_i = R_i - mean_j R_j`, and Eq. (8) multiplies that same scalar by the single final-sequence ratio:

```text
rho_i = pi_theta(y_i | x_i) / pi_theta_old(y_i | x_i)
```

There is no `k`-indexed policy ratio or other term that pairs `R_k` with the update at recurrent depth `k`. Because the external grader is nondifferentiable, every `R_k` influences the update only through the same aggregate trajectory advantage. Gradients can still propagate through the recurrent computation, but the objective does not localize each grader score to “the step that earned it.”

Thus LSRL provides **depth-informed trajectory reward shaping**, not depth-local policy-gradient credit in the sense asserted by the page’s definition.

## API-call count is also misstated

Section 3.2.2 says the **Process Grader is invoked twice per trajectory**, with each unified prompt assessing all `r` decoded snapshots; the two calls are averaged. A separate Final Answer Checker evaluates the completed solution. Section 3.3.1 confirms that each cached state is decoded, so there are `r` autoregressive decodes—but not `r` process-grader API calls.

For the reported `r = 8` setup, the paper describes eight snapshot decodes and two joint process-grader invocations (plus the separate final-answer check), rather than one grader invocation per depth.

## Consequence

The current text simultaneously overstates how precisely LSRL assigns credit and overstates its external-grader request count. Both affect the page’s comparison of LSRL with RLTT and the claimed cost of process supervision.

The reported 500-problem training set and +4.27 GSM8K result are accurate.

## Suggested correction

A compact replacement would be:

> LSRL decodes each cached Huginn depth and asks GPT-4.1-nano to score the snapshots. It discounts and sums the depth scores, mixes that scalar with final correctness (0.3/0.7), and uses the resulting single trajectory advantage in GRPO. This is depth-informed reward shaping rather than a depth-indexed policy-gradient objective. Its reported training workflow performs `r` autoregressive snapshot decodes and two joint process-grader calls per trajectory, plus a final-answer check—not one grader API call per depth.

The bibliography should likewise say “depth-scored, trajectory-aggregated reward” rather than implying that GRPO contains a separate reward/credit term for each depth.

**Comment by microprediction** [comment]
Correct on both counts, fixed in 1185867.

Verified against the EMNLP Findings paper. The per-depth scores go into R_proc = sum of gamma^(k-1) R_k with gamma 0.99, then R_tot combines that with final-answer correctness, then A_i = R_i - mean, and the GRPO ratio is a single whole-sequence quantity. No depth-indexed term survives, so it is reward shaping rather than dense per-depth credit.

And section 3.2.2 says the Process Grader is "invoked twice per trajectory (averaging two independent calls to reduce grader variance)", with one unified prompt covering all r snapshots. So the cost claim of an API call per depth was wrong as well. The page now says a decode per depth plus two grader calls per trajectory.
